File: src/api/issueStatus.ts
/**
 * 问题状态流转 API 封装
 * 基于 Vue3 + TypeScript + Axios
 */
import axios, { type AxiosResponse, type AxiosError } from 'axios'
import type { IssueStatus } from '@/types/issue'

// ============================================================================
// API 基础配置
// ============================================================================

const API_BASE_URL = '/api'

// 创建 axios 实例（复用 issue.ts 的配置）
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json'
  },
  timeout: 10000
})

// ============================================================================
// 响应拦截器
// ============================================================================

/**
 * 响应拦截器：统一处理响应数据
 */
apiClient.interceptors.response.use(
  (response) => {
    return response
  },
  (error: AxiosError) => {
    let errorMessage = '请求失败，请稍后重试'

    if (error.response) {
      const status = error.response.status
      const data: any = error.response.data

      switch (status) {
        case 400:
          errorMessage = data?.detail || data?.message || '请求参数错误'
          break
        case 401:
          errorMessage = '未授权，请登录后重试'
          break
        case 403:
          errorMessage = '无权访问该资源'
          break
        case 404:
          errorMessage = '请求的资源不存在'
          break
        case 500:
          errorMessage = '服务器内部错误'
          break
        default:
          errorMessage = data?.detail || data?.message || `请求失败 (${status})`
      }
    } else if (error.request) {
      errorMessage = '网络错误，请检查网络连接'
    } else {
      errorMessage = error.message || '未知错误'
    }

    return Promise.reject(new Error(errorMessage))
  }
)

// ============================================================================
// 类型定义
// ============================================================================

/**
 * 状态变更请求接口
 */
export interface IssueStatusChangeRequest {
  /** 目标状态 */
  to_status: IssueStatus
  /** 变更原因（可选） */
  reason?: string
}

/**
 * 状态变更响应接口
 */
export interface IssueStatusChangeResponse {
  /** 操作消息 */
  message: string
  /** 问题单号 */
  issue_no?: string
}

/**
 * 状态日志记录接口
 */
export interface IssueStatusLog {
  /** 日志 ID */
  id: number
  /** 问题 ID */
  issue_id: number
  /** 变更前状态 */
  from_status: string
  /** 变更后状态 */
  to_status: string
  /** 操作人 */
  operator: string
  /** 操作人角色（可选） */
  operator_role?: string
  /** 变更原因（可选） */
  reason?: string
  /** 创建时间 */
  created_at: string
}

/**
 * 状态日志列表响应接口
 */
export interface IssueStatusLogListResponse {
  /** 总记录数 */
  total: number
  /** 日志列表 */
  items: IssueStatusLog[]
}

/**
 * 状态流转规则接口（可选，用于前端校验）
 */
export interface StatusTransitionRule {
  /** 当前状态 */
  from: IssueStatus
  /** 允许的目标状态列表 */
  allowedTo: IssueStatus[]
  /** 流转说明 */
  description: string
}

// ============================================================================
// 状态流转规则常量（与后端保持一致）
// ============================================================================

/**
 * 状态流转规则映射
 * key=当前状态，value=允许流转的目标状态列表
 */
export const STATUS_TRANSITION_RULES: Record<IssueStatus, IssueStatus[]> = {
  '新建': ['处理中', '已关闭'],
  '处理中': ['已解决', '新建'],
  '已解决': ['已关闭', '处理中'],
  '已关闭': ['处理中'],
  'open': ['resolved', 'closed'],
  'resolved': ['closed', 'open'],
  'closed': ['open']
}

/**
 * 状态流转说明映射
 */
export const STATUS_TRANSITION_DESCRIPTION: Record<string, string> = {
  '新建 - 处理中': '问题开始处理',
  '新建 - 已关闭': '问题直接关闭（无需处理）',
  '处理中 - 已解决': '问题已解决',
  '处理中 - 新建': '问题重新打开',
  '已解决 - 已关闭': '问题确认关闭',
  '已解决 - 处理中': '问题重新处理',
  '已关闭 - 处理中': '问题重新激活'
}

// ============================================================================
// 工具函数
// ============================================================================

/**
 * 校验状态流转是否合法
 * @param fromStatus 当前状态
 * @param toStatus 目标状态
 * @returns { valid: 是否合法，message: 错误信息/说明 }
 */
export function validateStatusTransition(
  fromStatus: IssueStatus,
  toStatus: IssueStatus
): { valid: boolean; message: string } {
  // 状态相同，不允许
  if (fromStatus === toStatus) {
    return { valid: false, message: '目标状态与当前状态相同，无需变更' }
  }

  // 检查流转规则
  const allowedTransitions = STATUS_TRANSITION_RULES[fromStatus]
  if (!allowedTransitions || !allowedTransitions.includes(toStatus)) {
    return {
      valid: false,
      message: `不允许从状态'${fromStatus}'流转到'${toStatus}'，允许的目标状态：${allowedTransitions?.join('、') || '无'}`
    }
  }

  // 获取流转说明
  const transitionDesc = STATUS_TRANSITION_DESCRIPTION[`${fromStatus} - ${toStatus}`] || '状态变更'
  return { valid: true, message: transitionDesc }
}

/**
 * 获取状态流转说明
 * @param fromStatus 当前状态
 * @param toStatus 目标状态
 * @returns 流转说明
 */
export function getStatusTransitionDescription(
  fromStatus: IssueStatus,
  toStatus: IssueStatus
): string {
  const result = validateStatusTransition(fromStatus, toStatus)
  return result.message
}

// ============================================================================
// API 方法
// ============================================================================

/**
 * 变更问题状态
 * @param issueId 问题 ID
 * @param data 状态变更请求数据
 * @param operator 操作人（可选，默认 system）
 * @returns 变更结果
 */
export const changeIssueStatus = async (
  issueId: number,
  data: IssueStatusChangeRequest,
  operator?: string
): Promise<IssueStatusChangeResponse> => {
  const response: AxiosResponse<IssueStatusChangeResponse> = await apiClient.post(
    `/api/issues/${issueId}/status`,
    data,
    {
      params: {
        operator: operator || 'system'
      }
    }
  )
  return response.data
}

/**
 * 获取问题状态变更日志
 * @param issueId 问题 ID
 * @param limit 返回记录数量限制（可选，默认 50，最大 500）
 * @returns 状态日志列表
 */
export const getStatusLogs = async (
  issueId: number,
  limit?: number
): Promise<IssueStatusLogListResponse> => {
  const params: Record<string, any> = {}
  if (limit !== undefined) {
    params.limit = limit
  }

  const response: AxiosResponse<IssueStatusLogListResponse> = await apiClient.get(
    `/api/issues/${issueId}/status-logs`,
    { params }
  )
  return response.data
}

// ============================================================================
// 默认导出
// ============================================================================

export default {
  changeIssueStatus,
  getStatusLogs,
  validateStatusTransition,
  getStatusTransitionDescription,
  STATUS_TRANSITION_RULES,
  STATUS_TRANSITION_DESCRIPTION
}

File: backend/api/issue_status.py
"""
问题状态流转 API 接口模块
AutoCraft 问题跟踪系统 - Issue 状态变更 RESTful API
"""
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field

from database import get_db
from crud.issue import get_issue_by_id, update_issue
from crud.issue_status_log import create_status_log
from schemas.issue import IssueStatus, MessageResponse

router = APIRouter(prefix="/issues", tags=["问题状态管理"])


# ============================================================================
# 状态流转规则定义
# ============================================================================

# 状态流转规则：key=当前状态，value=允许流转的目标状态列表
STATUS_TRANSITION_RULES = {
    "新建": ["处理中", "已关闭"],
    "处理中": ["已解决", "新建"],
    "已解决": ["已关闭", "处理中"],
    "已关闭": ["处理中"],
}

# 状态流转说明
STATUS_TRANSITION_DESCRIPTION = {
    ("新建", "处理中"): "问题开始处理",
    ("新建", "已关闭"): "问题直接关闭（无需处理）",
    ("处理中", "已解决"): "问题已解决",
    ("处理中", "新建"): "问题重新打开",
    ("已解决", "已关闭"): "问题确认关闭",
    ("已解决", "处理中"): "问题重新处理",
    ("已关闭", "处理中"): "问题重新激活",
}


def validate_status_transition(from_status: str, to_status: str) -> tuple[bool, str]:
    """
    校验状态流转是否合法
    
    Args:
        from_status: 当前状态
        to_status: 目标状态
    
    Returns:
        (是否合法，错误信息/说明)
    """
    # 状态相同，不允许
    if from_status == to_status:
        return False, "目标状态与当前状态相同，无需变更"
    
    # 检查流转规则
    allowed_transitions = STATUS_TRANSITION_RULES.get(from_status, [])
    if to_status not in allowed_transitions:
        return False, f"不允许从状态'{from_status}'流转到'{to_status}'，允许的目标状态：{allowed_transitions}"
    
    # 获取流转说明
    transition_desc = STATUS_TRANSITION_DESCRIPTION.get((from_status, to_status), "状态变更")
    return True, transition_desc


# ============================================================================
# 请求/响应模型
# ============================================================================

class IssueStatusChangeRequest(BaseModel):
    """问题状态变更请求模型"""
    to_status: IssueStatus = Field(..., description="目标状态")
    reason: Optional[str] = Field(None, description="变更原因")


class IssueStatusLogResponse(BaseModel):
    """状态变更日志响应模型"""
    id: int = Field(..., description="日志 ID")
    issue_id: int = Field(..., description="Issue ID")
    from_status: str = Field(..., description="变更前状态")
    to_status: str = Field(..., description="变更后状态")
    operator: str = Field(..., description="操作人")
    operator_role: Optional[str] = Field(None, description="操作人角色")
    reason: Optional[str] = Field(None, description="变更原因")
    created_at: datetime = Field(..., description="创建时间")
    
    model_config = {"from_attributes": True}


class IssueStatusLogListResponse(BaseModel):
    """状态变更日志列表响应模型"""
    total: int = Field(..., description="总记录数")
    items: List[IssueStatusLogResponse] = Field(..., description="日志列表")


# ============================================================================
# API 接口
# ============================================================================

@router.post("/{issue_id}/status", response_model=MessageResponse)
def change_issue_status(
    issue_id: int,
    request: IssueStatusChangeRequest,
    db: Session = Depends(get_db),
    operator: Optional[str] = Query(default="system", description="操作人")
):
    """
    变更问题状态
    
    - 校验状态流转规则（新建→处理中→已解决→已关闭）
    - 自动记录状态变更日志
    - 支持指定操作人（默认 system）
    
    状态流转规则：
    - 新建 → 处理中、已关闭
    - 处理中 → 已解决、新建
    - 已解决 → 已关闭、处理中
    - 已关闭 → 处理中
    """
    # 1. 查询 Issue
    issue = get_issue_by_id(db, issue_id)
    if not issue:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"问题不存在：{issue_id}"
        )
    
    # 2. 校验状态流转
    from_status = issue.status
    to_status = request.to_status.value if hasattr(request.to_status, 'value') else str(request.to_status)
    
    is_valid, message = validate_status_transition(from_status, to_status)
    if not is_valid:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=message
        )
    
    # 3. 更新 Issue 状态
    try:
        from schemas.issue import IssueUpdate
        issue_update = IssueUpdate(status=to_status)
        updated_issue = update_issue(db, issue, issue_update)
        
        # 4. 记录状态变更日志
        create_status_log(
            db=db,
            issue_id=issue_id,
            from_status=from_status,
            to_status=to_status,
            operator=operator,
            reason=request.reason
        )
        
        return MessageResponse(
            message=f"问题状态已更新：{from_status} → {to_status}",
            issue_no=updated_issue.issue_no
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"状态变更失败：{str(e)}"
        )


@router.get("/{issue_id}/status-logs", response_model=IssueStatusLogListResponse)
def get_issue_status_logs(
    issue_id: int,
    limit: Optional[int] = Query(50, ge=1, le=500, description="返回记录数量限制"),
    db: Session = Depends(get_db)
):
    """
    获取问题状态变更日志
    
    - 返回指定 Issue 的状态变更历史
    - 按创建时间倒序排列
    - 默认返回最近 50 条记录
    """
    from crud.issue_status_log import get_status_logs_by_issue_id
    from schemas.issue_status_log import IssueStatusLogResponse
    
    issue = get_issue_by_id(db, issue_id)
    if not issue:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"问题不存在：{issue_id}"
        )
    
    try:
        logs = get_status_logs_by_issue_id(db, issue_id, limit=limit)
        return IssueStatusLogListResponse(
            total=len(logs),
            items=[IssueStatusLogResponse.model_validate(log) for log in logs]
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"获取状态日志失败：{str(e)}"
        )
